blog/_posts/2022-08-08-managing-tunnels-with-argocd.md

```diff
@@ -0,0 +1,236 @@
+  --controller-namespace=default \
+  --format yaml \
+  -f postgresql-customer1-token-secret.yaml > postgresql-customer1-token-sealed.yaml
+```
+
+All `SealedSecret` resources can be put in to the templates folder so that they can be deployed with the Helm chart.
+```
+# Example umbrella chart
+├── Chart.yaml
+├── templates
+│   └── postgresql-customer1-token-sealed.yaml
+└── values-postgresql-customer1.yaml
+```
+
+## Run multiple servers
+The steps described in the previous section can be repeated for each additional tunnel server that you want to run.
+
+Create an additional secret and add it to the chart:
+```bash
+# Generate a random token
+export TOKEN=$(head -c 16 /dev/random | shasum|cut -d" " -f1)
+
+# Create a yaml-encoded Secret without adding it to the cluster
+kubectl create secret generic mysql-customer2-token \
+  --dry-run=client \
+  --from-literal=token=$TOKEN \
+  -o yaml \
+  > mysql-customer2-token-secret.yaml
+
+# Create a SealedSecret
+kubeseal \
+  --controller-name=sealed-secrets \
+  --controller-namespace=default \
+  --format yaml \
+  -f mysql-customer2-token-secret.yaml > mysql-customer2-token-sealed.yaml
+```
+
+Add a new values file to the chart:
+```yaml
+# values-mysql-customer2.yaml
+inlets-pro:
+  tokenSecretName: mysql-customer2-token
+  ingress:
+    domain: mysql-customer2.example.com
+
+  dataPlane:
+    type: ClusterIP
+    ports:
+    - targetPort: 3306
+      protocol: TCP
+      name: mysql
+      port: 3306
+```
+
+Create a new Argo CD app:
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: mysql-customer2-tunnel
+  namespace: argocd
+spec:
+  project: default
+  source:
+    repoURL: https://github.com/welteki/inlets-server-argocd-demo.git
+    targetRevision: HEAD
+    path: charts/inlets
+    helm:
+        valueFiles:
+            - values-mysql-customer2.yaml
+  destination:
+    server: "https://kubernetes.default.svc"
+    namespace: default
+```
+
+Notice that the `spec.helm.valueFiles` field now references the `values.yaml` file for the second server.
+
+## Automate things further
+The pattern we described in the previous section works great of you only have to manage a couple of tunnel servers, but what if you have to manage dozens of servers? It can quickly become a mess to manage all the Argo CD Application resources. Luckily Argo CD provides a tool to automate the generation of Applications called [ApplicationSet](https://argo-cd.readthedocs.io/en/stable/user-guide/application-set/).
+
+The Application Set controller is deployed alongside Argo CD by default. Unlike an Argo CD Application resource, which deploys applications from a single source repository to a single destination cluster, the ApplicationSet CRD consists of [Generators](https://argocd-applicationset.readthedocs.io/en/stable/Generators/) and an Application template. A generator generates key/value parameters that are substituted in the Application template to render multiple applications.
```

Comment:
Is this App of Apps, or unrelated?

Reply:
This is a more generic solution that can be used to solve some of the limitations of the app of apps pattern.

I added a reference to the app of apps pattern at the end of the previous section.